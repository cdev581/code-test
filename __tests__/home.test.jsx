import "@testing-library/jest-dom";
import { render, screen, waitFor } from "@testing-library/react";
import { NextIntlClientProvider } from "next-intl";
import Home from "../src/app/page";
import hotelsData from "../src/data/hotels.json";

jest.mock("../src/app/components/loader/PlaneLoader.js", () => jest.fn().mockImplementation(() => {}));

beforeEach(() => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve(hotelsData),
    })
  )
});

afterEach(() => {
  jest.restoreAllMocks() // Clean up after each test
});

describe("Home", () => {
  const messages = require(`../messages/en.json`);
 
  it("renders homepage unchanged", async () => {
    const { container } = render(
      <NextIntlClientProvider messages={messages} locale="en">
        <Home />
      </NextIntlClientProvider>
    );
  
    await waitFor(() => 
      expect(container).toMatchSnapshot()
    );
  });

  it("renders a heading", async () => {
    const { container } = render(
      <NextIntlClientProvider messages={messages} locale="en">
        <Home />
      </NextIntlClientProvider>
    );
 
    const heading = screen.getByRole("heading", { level: 1 });
 
    await waitFor(() => 
      expect(heading).toBeInTheDocument()
    )
  });
});