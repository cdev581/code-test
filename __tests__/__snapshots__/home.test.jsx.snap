// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders homepage unchanged 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="container content"
    >
      <a
        class="logo"
        href="https://qantas.com"
      >
        <img
          alt="Qantas"
          data-nimg="1"
          decoding="async"
          height="30"
          src="/_next/image?url=%2Flogo.png&w=384&q=75"
          srcset="/_next/image?url=%2Flogo.png&w=256&q=75 1x, /_next/image?url=%2Flogo.png&w=384&q=75 2x"
          style="color: transparent;"
          width="152"
        />
      </a>
      <a
        class="github"
        href="https://github.com/cdev581/code-test"
      >
        <img
          alt="Github Repo"
          data-nimg="1"
          decoding="async"
          height="30"
          src="/_next/image?url=%2Fgithub-mark.png&w=64&q=75"
          srcset="/_next/image?url=%2Fgithub-mark.png&w=32&q=75 1x, /_next/image?url=%2Fgithub-mark.png&w=64&q=75 2x"
          style="color: transparent;"
          width="30"
        />
      </a>
    </div>
  </header>
  <div
    class="panel"
  >
    <section
      class="container"
    >
      <div
        class="text"
      >
        <h1>
          Frontend Code Test
        </h1>
        <p
          class="sub"
        >
          5th of February, 2025
        </p>
        <div
          class="columns"
        >
          <div
            class="column"
          >
            <p>
              This website is a coding test created for the purpose of a job application with Qantas Group. It is not an official website and is not associated with, endorsed by, or affiliated with Qantas Group. All branding and content is used solely for demonstration purposes.
            </p>
            <p>
              Thank you for taking the time to review my application and offering me the opportunity to participate. Immediately below is a React implementation of the results component that was depicted in the mock design. Below that, I have also included a discussion of my solution approach, as well as the rationale behind some technical decisions.
            </p>
          </div>
          <div
            class="column"
          >
            <p>
              For this project I have utilised 
              <strong>
                Next.js
              </strong>
               with 
              <strong>
                App Router, React
              </strong>
               and 
              <strong>
                TypeScript
              </strong>
              . Styles have been applied using the CSS preprocessor 
              <strong>
                SCSS
              </strong>
              , and 
              <strong>
                svgr
              </strong>
               is imported as an SVG loader. I have added localisation using 
              <strong>
                next-intl
              </strong>
               and have written some basic sample tests using 
              <strong>
                Jest
              </strong>
              . Finally I have enhanced the loading message with 
              <strong>
                WebGL/three.js
              </strong>
              , using a plane model I created in Blender (loosely based on a 787).
            </p>
            <p>
              Thank you again for your consideration. Please don't hesitate to reach out if you require any other information from me regarding my application.
            </p>
          </div>
        </div>
      </div>
    </section>
    <div
      class="container-thin"
    >
      <section
        class="showcase"
      >
        <section>
          <div
            aria-labelledby="status"
            id="results"
            role="region"
          >
            <header
              class="header"
            >
              <img
                alt="Qantas"
                data-nimg="1"
                decoding="async"
                height="30"
                src="/_next/image?url=%2Flogo.png&w=384&q=75"
                srcset="/_next/image?url=%2Flogo.png&w=256&q=75 1x, /_next/image?url=%2Flogo.png&w=384&q=75 2x"
                style="color: transparent;"
                width="152"
              />
              <div
                class="refine"
              >
                <h2
                  aria-live="polite"
                  class="title"
                  id="status"
                >
                  <strong>
                    Loading Results
                  </strong>
                </h2>
                <div
                  class="sort"
                >
                  <label
                    class="sort-label"
                    for="hotels-results-sort"
                  >
                    Sort by
                  </label>
                  <select
                    aria-controls="results"
                    id="hotels-results-sort"
                  >
                    <option
                      value="price_asc"
                    >
                      Price low-high
                    </option>
                    <option
                      value="price_desc"
                    >
                      Price high-low
                    </option>
                  </select>
                </div>
              </div>
            </header>
            <div
              class="results"
            >
              <div
                class="
              loader
              
            "
              >
                <canvas
                  class="canvas"
                  id="plane-canvas"
                />
              </div>
              <ul
                class="list hidden"
              />
            </div>
          </div>
        </section>
      </section>
    </div>
  </div>
  <section
    class="panel callout"
  >
    <div
      class="container-thin"
    >
      <div
        class="text"
      >
        <h2>
          Design Analysis
        </h2>
        <div
          class="columns"
        >
          <div
            class="column"
          >
            <p>
              After examining the mock design, I believe 
              <strong>
                Helvetica
              </strong>
               or 
              <strong>
                Helvetica Neue
              </strong>
               is the primary font used. Since I do not have a web licence for Helvetica, I have selected the Google font 
              <strong>
                Inter
              </strong>
               as an alternative. I did consider providing a font stack including Helvetica when it was available, but I have instead prioritised a consistent visual experience.
            </p>
            <p>
              No mobile or tablet layouts were provided, so I have created my own based on common design patterns. If the viewport is thin enough, the content within each hotel item will stack and appear as a card.
            </p>
            <p>
              The images in the design display at 
              <strong>
                150px
              </strong>
               by 
              <strong>
                125px
              </strong>
              . However, the mock JSON endpoint selects images from unsplash at a size of 
              <strong>
                145px
              </strong>
               by 
              <strong>
                125px
              </strong>
              . To account for this, 
              <strong>
                object-fit: cover
              </strong>
               has been applied so the image does not become distorted. Since the JSON was provided as part of the test materials I have not modified it, but under normal circumstances I would suggest requesting a larger image size to account for devices with higher DPIs.
            </p>
            <p>
               Finally, I have examined the arrangement of the result items carefully in order to understand the intent of the design. This helped me make decisions on how to structure the HTML and CSS.
            </p>
            <ul>
              <li>
                None of the text wraps.
              </li>
              <li>
                Each result item is always the same height.
              </li>
              <li>
                Each piece of information within result items is always in the same position if present.
              </li>
            </ul>
            <p>
              From the above points it is clear that the designer values vertical rhythm and a consistent user experience. I have retained these principals in the implementation, including 
              <strong>
                text-overflow: ellipsis
              </strong>
               and maintaining negative space when some elements are missing. There are some accessibility trade-offs with this approach that will be discussed in its own section.
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section
    class="panel"
  >
    <div
      class="container-thin"
    >
      <div
        class="text"
      >
        <h2>
          Architecture
        </h2>
        <div
          class="columns"
        >
          <div
            class="column"
          >
            <p>
              I have created a mock endpoint using NextResponse, which reads the JSON file from the system. The results are sorted based on a provided query param, then displayed by a client component. I have created a generic component 
              <strong>
                Result List
              </strong>
               which handles retrieving the response and passes it to it's children. I also wrote a custom hook which abstracts the fetch call, and provides the response data and a status flag.
            </p>
            <p>
              The loading animation was created using three.js, the canvas itself is in React, but is memoized to prevent it from being re-rendered. It's not ideal to load such a large model in order to display a loading icon, but since this is just a demo project I decided to adhere to the 
              <strong>
                Rule of Cool
              </strong>
              . The load time is artificially inflated server side with a setTimeout in order to show the loading animation.
            </p>
            <p>
              The ratings component has been accomplished using an SVG with a mask embedded in the file. There is a rectangle underneath affected by the mask which has its length updated based on the rating using TypeScript. Potentially the colours could also be changed using stylesheets since the SVG's markup is loaded onto the page, not into an image component. One drawback here is that the mask circles/stars have gaps between them, so setting the fill length to the rating as a percentage of the total width may not align with the stars exactly.
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section
    class="panel callout"
  >
    <div
      class="container-thin"
    >
      <div
        class="text"
      >
        <h2>
          Accessibility
        </h2>
        <div
          class="columns"
        >
          <div
            class="column"
          >
            <p>
              The results component has several accessibility features built into it. These include but aren't limited to:
            </p>
            <ul>
              <li>
                <strong>
                  High colour contrast
                </strong>
                : I have slightly darkened some of the copy compared to the design to create a colour contrast greater than 4.5:1.
              </li>
              <li>
                <strong>
                  Visible focus outline
                </strong>
                : Using the focus-visible pseudo selector, a dark outline has been added to elements that may be interacted wtih.
              </li>
              <li>
                <strong>
                  Accessible anchor content
                </strong>
                : In most cases, the desired behaviour of result items is for the entire element to be clickable. However, wrapping the entire element in an anchor can provide difficulties for users of screen readers since it is too much content for a single anchor. To resolve this, only the heading is inside an anchor, and a simulated click event has been applied to the rest of the card.
              </li>
              <li>
                <strong>
                  Semantic HTML and ARIA attributes:
                </strong>
                 Where possible, relavent HTML tags and ARIA attributes have been used to aid users with accessibility peripherals. These include role='region', aria-live='polite', aria-labelledby, aria-controls, alt tags and others.
              </li>
            </ul>
            <p>
              Additionally, thought has been given to the scalibility of the design. There are three main ways a user can zoom on a webpage.
            </p>
            <ul>
              <li>
                <strong>
                  Pinch to zoom
                </strong>
                : Common on touch devices. Not ideal as the user may need to scroll horizontally to read the entire line length.
              </li>
              <li>
                <strong>
                  Setting browser zoom
                </strong>
                : Up-scales the entire webpage. This is most common on desktop browsers. Zooming too far can activate smaller breakpoints, further obscuring content.
              </li>
              <li>
                <strong>
                  Text-only scaling
                </strong>
                : Rather than scaling the entire document uniformly, only CSS metrics using font units will scale. This is commonly used on mobile devices, where the layout does not allow for uniform up-scaling.
              </li>
            </ul>
            <p>
              In this project, consideration has been given as to where to use scalable units such as 
              <strong>
                rem
              </strong>
               vs fixed units like 
              <strong>
                px
              </strong>
              . In the above design, text-overflow is used to prevent text from wrapping. However, this has the drawback of reducing the amount of copy visible to users who are using the text-only scaling. To account for this, I have specified the container's width using rem instead of px. Therefore when the user increases the text size, the container should also grow, giving more room for the text to fit.
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
